### PR TITLE
CW2 28 create modal on sponsor click

### PR DESCRIPTION
**Pull request by QuadAces**
### Why the changes are required?
Because we want a modal to show up when we click on a sponsor!
### Changes
- added a setstate to keep track of a modal
- added a setstate to keep track of modal information (yes I could use useReducer or I could just not)
- changed sponsor TS interface to have a description
- added modal component
### Screenshots
Screenshot 2024-06-20 at 8 47 43 pm
Screenshot 2024-06-20 at 8 47 54 pm
Screenshot 2024-06-20 at 8 48 07 pm
### Comments
May I say that the datastructure to keep track of our sponsors is awful. Since we have objects for each sponsor level (const diamond = {} and also have const gold = {}) I think this is really hard to work with as I had to manipulate each of the components mapped to show a modal separately. I also suggest adding a name to the sponsors datastructure field also.

Lastly, there are some errors with the source for some of the company logos and svgs. I understand Ming is working on this, just note that some svgs/images dont show on click for the modal.

### CODE REVIEW

File: frontend/src/components/Sponsors/sponsorModal.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { MouseEventHandler } from 'react';
```
Comment (derekxu04) [inline frontend/src/components/Sponsors/sponsorModal.tsx:1]: Can delete this line

File: frontend/src/components/Sponsors/sponsorModal.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import { sponsorInfo } from '../../../public/data/data';
+
+export default function SponsorModal(props: { sponsorInfo: sponsorInfo | null; setFalse: any }) {
```
Comment (derekxu04) [inline frontend/src/components/Sponsors/sponsorModal.tsx:4]: Can make the type of setFalse to be a function (something like `() => void` or something)

File: frontend/src/components/Sponsors/sponsorModal.tsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  return (
+    <div
+      className="fixed inset-0 flex items-center justify-center bg-black bg-opacity-50"
+      onClick={() => {
+        props.setFalse();
+      }}
+    >
+      <div className="bg-[#3977f8] mx-[10vw] py-10 rounded-xl overflow-auto flex flex-col items-center justify-center w-[800px]">
+        <a className="m-10" href={props.sponsorInfo.href}>
+          <img src={`./${props.sponsorInfo.svg}`} alt={props.sponsorInfo.alt} />
```
Comment (derekxu04) [inline frontend/src/components/Sponsors/sponsorModal.tsx:21]: Can we specify a width / height cause I noticed some logos are much bigger / smaller than others

File: frontend/src/components/Sponsors/sponsorModal.tsx [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        </a>
+        <h3 className="mx-10 py-10">{props.sponsorInfo.description}</h3>
+        <button
+          onClick={props.setFalse}
+          className="bg-white border text-[#3977F8] border-[#A7A6E5] text-lg rounded-xl w-[70%] xl:h-12 h-10"
+        >
+          close
```
Comment (derekxu04) [inline frontend/src/components/Sponsors/sponsorModal.tsx:28]: I think making this `Close` looks better

File: frontend/src/components/Sponsors/sponsorlinks.tsx
Comment (derekxu04) [inline frontend/src/components/Sponsors/sponsorlinks.tsx:1]: Can we rename this file `SponsorLinks.tsx` just to keep consistent

File: frontend/src/components/Sponsors/sponsorModal.tsx
Comment (derekxu04) [inline frontend/src/components/Sponsors/sponsorModal.tsx:1]: Can we rename this `SponsorModal.tsx` for consistency

File: frontend/src/components/Sponsors/sponsorlinks.tsx [hunk 5]
```diff
@@ -1,24 +1,44 @@
...
+              setShowModal(false);
+            }}
+          />
+        )}
         <div
           style={{ backgroundColor: `${background}` }}
           className="flex flex-wrap rounded-[1rem] pl-14 py-14 gap-16 items-center"
         >
           <h2 className="text-4xl font-black">Diamond Sponsors</h2>
           {diamondLinks.map((item, index) => {
```
Comment (derekxu04) [inline frontend/src/components/Sponsors/sponsorlinks.tsx:30]: Would be good for the logo to maybe expand slightly and for the cursor to become a pointer on hover to indicate it is clickable (applies to all logos)

**Review by derekxu04 — CHANGES_REQUESTED** [review]
Nice work, looks really good!

Just some changes:
- I've noted some changes above on specific lines
- Would be nice to either indicate that the logo is clickable in the modal, or to just have a "Learn more" button / text or something, otherwise it is not super obvious 

Comments:
- Do you know why some of the svgs don't show up?
- With regards to the format of `data.ts`, I don't see any way of avoiding having to split the sponsors into the three sections. If we want to avoid repeating the code to do with the modal, we can make it even more modular and make a `SponsorSection` component or something for each level, which would work I think, or maybe just write a function that does that for you without needing to make it its own entire new component
- Agree with adding the `name` property to each company, either you can do that or I'll can do that at some point

File: frontend/src/components/Sponsors/sponsorModal.tsx (hunk 2 above)
Comment (QuadAces) [reply frontend/src/components/Sponsors/sponsorModal.tsx:4]: yes sorry I don't know how to deal with TS to be honest, and I searched it up and always got some error unless I //@ts-ignored
I'll look into it.

File: frontend/src/components/Sponsors/sponsorlinks.tsx (hunk 5 above)
Comment (QuadAces) [reply frontend/src/components/Sponsors/sponsorlinks.tsx:30]: Yep for sure. I tried making everything responsive. I tested myself and found that when the logos are hovered mouse defaults to pointer anyway, could you show me what you mean?

**Comment by QuadAces** [comment]
Okay great, done pretty much everything. for data ts, I just propse that we have an object with 3 key values just to at least connect the data somehow e.g. {diamond sponsors: [], gold sponsors: []}
I obviously dont want to make things more complex than they should be, just think they should be grouped together. I've added everything else :P

**Review by derekxu04 — APPROVED** [review]

**Comment by derekxu04** [comment]
> Okay great, done pretty much everything. for data ts, I just propse that we have an object with 3 key values just to at least connect the data somehow e.g. {diamond sponsors: [], gold sponsors: []}

Yep that makes sense I'll do this in another ticket

**Pull request merged**